Quantize inputs with a session initialization phase. (#307)
ietf-rats-wg/draft-ietf-rats-corim: draft-ietf-rats-corim.md

  IANA.named-information: named-info

informative:

>   RFC6960:

Reviewer: ```suggestion
  RFC6960: ocsp
```

The appraisal procedure is divided into several logical phases for clarity.

+ **Phase 0**: Session setup

> During Phase 0, the Verifier collects its set of inputs from external sources that will be used for the remainder of the Appraisal Procedure to ensure the Procedure is deterministic.

Reviewer: The deterministic nature is ensured by the consistency of the appraisal algorithm.
```suggestion
During Phase 0, the Verifier collects its set of inputs from external sources that will be used for the remainder of the Appraisal Procedure.
```

Author: The appraisal algorithm is deterministic for the base CDDL, but profiles can define their own comparison algorithms that can be complex and have computational timeouts that lead to nondeterministic results (for example). Considering the SLSA level 4 language about hermeticity and reproducibility, we should have wording about best-effort reproducibility https://slsa.dev/spec/v0.1/requirements

If we consider a Verifier to be a "builder" of attestation results, then we can apply the same requirements whole sale.

Reviewer: Hmm, even with profile-specific comparison algorithms (or other kinds of custom behaviour), _successful_ ACS computation should be deterministic.

If computational timeouts, or other conditions that make the ACS "incomplete", happen, they must be made fully visible to the downstream appraisal policy.

Author: Suppose we have a profile that has unification variables that are allowed in conditions and in additions. The logic it employs for unification does not have principal types, so there are multiple incomparable correct solutions to a system of constraints. Which substitution it resolves on can be dependent on something nondeterministic like hash map iteration order. The resulting ACS is not consistently reproducible in this case, the way a compiler can choose which way it allocates registers or orders if/then/else basic blocks.

I'm not saying it's sensible, I'm saying it's not ruled out and it is very difficult to rule out. Thus a SHOULD be reproducible.

Author: We've already said that certain decisions can be make by policy. Are those decisions made strictly after phase 4? If we can say phases 2, 3, and 4 must be deterministic without imposing policy burden by the way we define semantics, then I'd certainly like to make that declaration.

I don't think profiles would be introducing randomized behavior. Non-deterministic isn't necessarily random. You can have "the same" outcome by means of equivalence but not syntactic equality. When we're talking about ACS construction we're making representation choices that may be more conducive to equal outputs than something that uses, say, hash maps or algebraic constructions that represent "the same object" in multiple ways.

Reviewer: > We've already said that certain decisions can be make by policy. Are those decisions made strictly after phase 4?

yes

> If we can say phases 2, 3, and 4 must be deterministic without imposing policy burden by the way we define semantics, then I'd certainly like to make that declaration.

👍 

> I don't think profiles would be introducing randomized behavior. Non-deterministic isn't necessarily random. You can have "the same" outcome by means of equivalence but not syntactic equality. When we're talking about ACS construction we're making representation choices that may be more conducive to equal outputs than something that uses, say, hash maps or algebraic constructions that represent "the same object" in multiple ways.

Got it.  Then:

> _"profiles MUST NOT introduce non-deterministic behaviour."_

: A structure that holds ECTs that have been appraised.
The ACS contains Attester state that has been authorized by Verifier processing and Appraisal Policy.

Appraisal Context:
: A structure that contains all state needed for performing the Appraisal Procedure.

Appraisal Policy:
: A description of the conditions that, if met, allow acceptance of Claims. Typically, the entity asserting a Claim should have knowledge, expertise, or context that gives credibility to the assertion. Appraisal Policy resolves which entities are credible and under what conditions.  See also "Appraisal Policy for Evidence" in {{-rats-arch}}.

Attestation Results Set (ARS):
: A structure that holds results of Appraisal and ECTs that are to be conveyed to a Relying Party.

Appraisal Session:

> : A structure that tracks all state that corresponds to a single request for attestation appraisal.

Author: Removed.

: A structure that holds ECTs that have been appraised.
The ACS contains Attester state that has been authorized by Verifier processing and Appraisal Policy.

Appraisal Context:
: A structure that contains all state needed for performing the Appraisal Procedure.

Appraisal Policy:
: A description of the conditions that, if met, allow acceptance of Claims. Typically, the entity asserting a Claim should have knowledge, expertise, or context that gives credibility to the assertion. Appraisal Policy resolves which entities are credible and under what conditions.  See also "Appraisal Policy for Evidence" in {{-rats-arch}}.

Attestation Results Set (ARS):
: A structure that holds results of Appraisal and ECTs that are to be conveyed to a Relying Party.

Appraisal Session:

> : A structure that tracks all state that corresponds to a single request for attestation appraisal. This includes the Appraisal Claims Set and Appraisal Context.

Reviewer 2: ```suggestion
: A structure that tracks all state that corresponds to a single request for attestation appraisal.
This includes the Appraisal Claims Set and Appraisal Context.
```
We're breaking lines on sentences right?

{::include cddl/intrep-ars.cddl}
~~~

## Session setup (Phase 0) {#sec-phase0}

The exchange of a request for attestation appraisal for a response of Attestation Results corresponds to a single Attestation Session.

During this setup phase, the Verifier populates its Appraisal Session with a consistent view of all its inputs to the Appraisal Procedure.
Inputs are various conceptual messages collected from Reference Value Providers, Endorsers, Verifier Owners, and Attesters.

> Conceptual messages may include Attestation Evidence, CoMID tags {{sec-comid}}, CoSWID tags {{-coswid}}, CoBOM tags {{sec-cobom}}, and static cryptographic validation key material (including raw public keys, root certificates, intermediate CA certificate chains, and Concise Trust Anchor Stores (CoTS) {{-ta-store}}.

Author: Yeah.. Ned's change to add parentheses didn't treat this one line consistently, and adding parens only for CoBOM seems weird, no?

> Conceptual messages may include Attestation Evidence, CoMID tags {{sec-comid}}, CoSWID tags {{-coswid}}, CoBOM tags {{sec-cobom}}, and static cryptographic validation key material (including raw public keys, root certificates, intermediate CA certificate chains, and Concise Trust Anchor Stores (CoTS) {{-ta-store}}.

Reviewer: > [...] adding parens only for CoBOM seems weird, no?

It makes sense to add parentheses because external refs are rendered differently from internal refs:

* `{{EXT}}` becomes `[EXT]`
* `{{INT}}` becomes `INT`

> Conceptual messages may include Attestation Evidence, CoMID tags {{sec-comid}}, CoSWID tags {{-coswid}}, CoBOM tags {{sec-cobom}}, and static cryptographic validation key material (including raw public keys, root certificates, intermediate CA certificate chains, and Concise Trust Anchor Stores (CoTS) {{-ta-store}}.

Author: Added to both cobom and comid references then.

> It is left to Verifier Policy to determine if input sources must use supply chain transparency constructs (see {{-scitt-arch}}) to track input provenance.

Author: Verifier Policy can be to follow regulations or not. I didn't want to add speculation about this spec going to regulators as a framework.

> It is left to Verifier Policy to determine if input sources must use supply chain transparency constructs (see {{-scitt-arch}}) to track input provenance.

Reviewer: Understood. The thing is 9334's definition of policy is:

> _"A set of rules that a Verifier uses to evaluate the validity of information about an Attester."_

which would make any regulatory aspects out of scope.

The time at which the Verifier evaluates certificate- and tag-validity is an input.
Once the Appraisal Session inputs are collected, no more may be added to the Appraisal Session apart from one exception.

> Certificate revocation status results may be collected during Phase 1, since there is no collective simultaneity of all responses.

Author: Sounds like you're saying I should allow for the wall clock input to also be collected multiple times during Phase 1? I'm just trying to temper all sources of external influence on the hermeticity of an appraisal.

> Certificate revocation status results may be collected during Phase 1, since there is no collective simultaneity of all responses.

Reviewer: I am just pointing out that creating an exception just for revocation checks to the general rule that _"Phase 0 locks all inputs"_ is probably unnecessary.  If we go without, the loss of functionality doesn't seem too big, given:
a) Temporal distance between Phases 0 and 1 is likely to be (very) short, and
b) Some clock skew tolerance will be in place anyway, making the boundaries between 0 and 1 even fuzzier.

> Certificate revocation status results may be collected during Phase 1, since there is no collective simultaneity of all responses.

Author: Moved the lock to be back at the end of Appraisal Context construction then.

> It is left to Verifier Policy to determine if or how to log the inputs used for a given Appraisal Session for optional use in Attestation Results.

Author: Do you want different text?

> It is left to Verifier Policy to determine if or how to log the inputs used for a given Appraisal Session for optional use in Attestation Results.

Reviewer: Yes, just a small tweak to acknowledge external factors other than the Verifier Owner's will may be at play.

> It is left to Verifier Policy to determine if or how to log the inputs used for a given Appraisal Session for optional use in Attestation Results.

Author: Added.

Note: Verifiers may rely on conveyance protocol–specific context to identify an Evidence source, which is the Evidence input oracle for appraisal.

## Input Validation and Transformation (Phase 1) {#sec-phase1}

In Phase 1 the Verifier constructs an Appraisal Context that will serve as the set of valid sources of information for the Appraisal Procedure.
The primary goal of this phase is to ensure that all necessary information is valid and available for subsequent processing.

> No inputs other than dynamically-fetched information about certificate revocation status (see [RFC6960]) may be collected in Phase 1.

Reviewer: ```suggestion
No inputs other than dynamically-fetched information about certificate revocation status (see {{?RFC6960}}) may be collected in Phase 1.
```

Author: Should I add RFC6960 to the normative frontmatter instead? I don't know this {{?ref}} syntax.

Reviewer: `{{?ref}}` automatically creates an informational reference entry for `ref`.
`{{!ref}}` automatically creates a normative reference entry for `ref`.

It's handy for GitHub's "suggested changes" because you don't have to fiddle with the front matter when you are miles deep in the scroll.

Reviewer: > Should I add RFC6960 to the normative frontmatter instead? 

I would make OCSP informative.  Using CRL distribution points is also an option.

Author: Done.

{::include cddl/intrep-ars.cddl}
~~~

### Internal Representation of Appraisal Session {#sec-ir-asession}

An Appraisal Session includes Verifier-specific session state as well as a collection of inputs to process.
Conceptual Messages are given explicit representation in the session.

~~~ cddl

> {::include cddl/intrep-asession.cddl}

Author: Removed appraisal session representation.

~~~

The session state is implementation-specific, but conceptual messages defined in this specification are specially represented.

## Appraisal Context Construction (Phase 1) {#sec-phase1}

In Phase 1 the Verifier constructs an Appraisal Context that will serve as the set of valid sources of information for the Appraisal Procedure.
The primary goal of this phase is to ensure that all necessary information is valid and available for subsequent processing.

~~~ cddl
{::include cddl/intrep-actx.cddl}
~~~

### Input Collection {#sec-phase1-collect}

> The exchange of a request for attestation appraisal for a response of Attestation Results corresponds to a single Attestation Session.

Author: It's not. The results are produced at the end of a session. An Appraisal Context is only for getting to the end of constructing the ACS. It's a bit matryoshka doll-ish.

During this setup phase, the Verifier populates its Appraisal Session with a consistent view of all its inputs to the Appraisal Procedure.
Inputs are various conceptual messages collected from Reference Value Providers, Endorsers, Verifier Owners, and Attesters.
Conceptual messages may include Attestation Evidence, CoMID tags ({{sec-comid}}), CoSWID tags {{-coswid}}, CoBOM tags ({{sec-cobom}}), Policy, and cryptographic validation key material (including raw public keys, root certificates, intermediate CA certificate chains, certificate revocation data (see {{-ocsp}} or {{Section 4.2.1.13 of -pkix-cert}}), and Concise Trust Anchor Stores (CoTS) {{-ta-store}}.
The clock time used for validity judgments and policy evaluation is an input.

How the Verifier collects its inputs is out of scope of this document.

> There will be some amount of CoRIMs and standalone tags available as inputs that make an `asession`:

Author: Okay, I'll remove the internal representation and leave the description to the prose of the Verifier Abstraction.

~~~ cddl
{::include cddl/intrep-asession.cddl}`
~~~

Initially all inputs are in `cms` if interpreted by this specification, or `extra` if not.

> It is left to Verifier Policy to determine if input sources must use supply chain transparency constructs (see {{-scitt-arch}}) to track input provenance.

Author: I lost track of what this line used to refer to.

> The selection process MUST yield at least one usable conceptual message.
> Dionna: I don't think this should be a requirement since the evidence combined with Verifier policy should be enough.

#### Tag Extraction and Validation

From the selected CoRIMs, the tags it contains also go through selection and validation.

The Verifier MUST discard all tags which are not syntactically or semantically valid.
The Verifier MUST discard all tags not within their validity period.

In particular, any cross-referenced triples (e.g., CoMID-CoSWID linking triples and CoBOM-listed tags) MUST reference valid entries of the Appraisal Context.
A CoMID's `linked-tags` field is able to express cyclic references, but cyclic references that include `/ tag-rel / 1: / supplements / 0` or `/ tag-rel / 1: / replaces/  1` are invalid.
Unless otherwise specified by a profile that extends `$tag-rel-type-choice`, all cyclic tag references are invalid.

#### Tag activation by CoBOM

A tag's activity is determined by the flowchart in figure {#fig-tag-activity}

flowchart TD
    A[unknown] --> C{use_cobom}
    C -->|true| D{is_listed}
    C -->|false| E[active]
    D -->|true| F[active]

>     D -->|false| G[inactive]

Author: The referenced issue does not appear to have solved the problem. Do you know if this change would work?

>     D -->|false| G[inactive]

Reviewer: I believe it wouldn't because of point 3 (i.e., `mmdc` not producing conformant SVG)

{::include cddl/intrep-ars.cddl}
~~~

> ## Session setup (Phase 0) {#sec-phase0}

Reviewer 2: Why change from phase 1 to phase 0?

It is left to Verifier Policy to determine if input sources must use supply chain transparency constructs (see {{-scitt-arch}}) to track input provenance.

The time at which the Verifier evaluates certificate- and tag-validity is an input.
Once the Appraisal Session inputs are collected, no more may be added to the Appraisal Session apart from one exception.
Certificate revocation status results may be collected during Phase 1, since there is no collective simultaneity of all responses.

It is left to Verifier Policy to determine if or how to log the inputs used for a given Appraisal Session for optional use in Attestation Results.

Note: Verifiers may rely on conveyance protocol–specific context to identify an Evidence source, which is the Evidence input oracle for appraisal.

## Input Validation and Transformation (Phase 1) {#sec-phase1}

In Phase 1 the Verifier constructs an Appraisal Context that will serve as the set of valid sources of information for the Appraisal Procedure.
The primary goal of this phase is to ensure that all necessary information is valid and available for subsequent processing.

> No inputs other than dynamically-fetched information about certificate revocation status (see {{RFC6960}}, and {{Section 4.2.1.13 of -pkix-cert}}) may be collected in Phase 1.

Reviewer 2: I thought the previous text was better / orthogonal to the suggested text.
```suggestion
After the Appraisal Context is initialized, appraisal processing is allowed to complete before new inputs are considered.
```
I thought the previous text was better / orthogonal to the suggested text.

Author: The “allowed” worries me, since we don’t require monotonic matching conditions for profiles and mid-appraisal additions of tags can lead to inconsistent ACS results.

Reviewer: Party pooper mode on. (For different reasons) I don't like either 🤣

Author: I've added a hermeticity subsection to describe this without any need for operational suggestions like having inputs be "held back".

> No inputs other than dynamically-fetched information about certificate revocation status (see {{RFC6960}}, and {{Section 4.2.1.13 of -pkix-cert}}) may be collected in Phase 1.

Reviewer: ```suggestion
No inputs other than dynamically-fetched information about certificate revocation status (see {{-ocsp}}, and {{Section 4.2.1.13 of -pkix-cert}}) may be collected in Phase 1.
```

{::include cddl/intrep-ars.cddl}
~~~

## Appraisal Context Construction (Phase 1) {#sec-phase1}

In Phase 1 the Verifier constructs an Appraisal Context that contains valid inputs to the Appraisal Procedure.
The Appraisal Context is available to the Appraisal Procedure throughout the various phases.

~~~ cddl
{::include cddl/intrep-actx.cddl}
~~~

### Input Collection {#sec-phase1-collect}

> The exchange of a request for attestation appraisal for a response of Attestation Results corresponds to a single Attestation Session.

Reviewer 3: ```suggestion
A single attestation session is a period when a attestation appraisal process is initiated upon Verifier receiving Evidence from Attester or Relying Party and generating the results of the Appraisal.
```

Inputs that are not within their validity period, or that cannot be associated with an authenticated and authorized source MUST be discarded from the session.

Any input that has been secured by a cryptographic mechanism, such as a signature, that fails validation MUST be discarded from the session.

Selected tags are added to `/ select /` by mapping their tag identifier to a `tag-state` that is a pair of the tag and `activity: unknown`.

Other selection criteria MAY be applied.
For example, if the Evidence format is known in advance, CoRIMs using a profile that is not understood by a Verifier for that Evidence format MAY be discarded.
Selection policies model the trust relationships between the Verifier Owner and the relevant suppliers, and are out of the scope of the present document.
For example, a composite device ({{Section 3.3 of -rats-arch}}) is likely to be fully described by multiple CoRIMs, each signed by a different supplier.
In such case, the Verifier Owner may instruct the Verifier to discard tags activated by supplier CoBOMs that are not also activated by the trusted integrator.

> > The selection process MUST yield at least one usable conceptual message.

Reviewer 3: Need to discuss this..

The selection process MUST yield at least one usable tag.

Selected CoRIMs are transformed into an internal representation (see {{sec-phase1-trans}}).
If CoBOMs are required by policy, the selected tags are considered *inactive*.

> An *inactive* tag MUST NOT be used during the Appraisal Procedure.

Reviewer: I find this way of presenting activation/disactivation a bit puzzling.

I see the flow from undecided/unknown to either active or inactive.
It'd be probably easier to grok if we introduced an initial "unknown" state and describe the decision tree as:

```mermaid
flowchart TD
    A[unknown] --> C{use_cobom}
    C -->|true| D{is_listed}
    C -->|false| E[active]
    D -->|true| F[active]
    D -->|false| G[inactive]
```

Reviewer 2: > A diagram like this would be helpful IMO.

But maybe the diagram should be backed out of this PR until the tooling issues can be addressed. It's difficult to suggest changes when it doesn't build.

If CoBOMs are not required by policy, the selected tags are considered *active* and thus MUST be used during the Appraisal Procedure.

Later stages will further select the CoRIMs appropriate to the Evidence Appraisal stage.

#### Tags Extraction and Validation

The Verifier chooses tags from the selected CoRIMs—including CoMID, CoSWID, CoBOM, and CoTS.

The Verifier MUST discard all tags which are not syntactically and semantically valid.
In particular, any cross-referenced triples (e.g., CoMID-CoSWID linking triples and CoBOM-listed tags) MUST be successfully resolved.

#### Tag activation by CoBOM

> This step can be skipped of all tags are implicitly active.

Reviewer: Doesn't read right to this non-native speeker. Did you mean s/of/if/? Or something else?

Regardless, I prefer it stupid and explicit, as it was before :-)
Unless you see a case that would not be covered using the old phrasing.

Author: Yes s/of/if/, typo.

CoBOMs which are not within their validity period are discarded.

> The Verifier processes all CoBOMs that are valid at the point in time of Evidence Appraisal.

Author: Removed.

The selection process MUST yield at least one usable tag.

Selected CoRIMs are transformed into an internal representation (see {{sec-phase1-trans}}).
If CoBOMs are required by policy, the selected tags are considered *inactive*.
An *inactive* tag MUST NOT be used during the Appraisal Procedure.
If CoBOMs are not required by policy, the selected tags are considered *active* and thus MUST be used during the Appraisal Procedure.

Later stages will further select the CoRIMs appropriate to the Evidence Appraisal stage.

#### Tags Extraction and Validation

> The Verifier chooses tags from the selected CoRIMs—including CoMID, CoSWID, CoBOM, and CoTS.

Reviewer: ```suggestion
The Verifier chooses tags from the selected CoRIMs, including CoMID, CoSWID, CoBOM, and CoTS.
```

Reviewer: ah, got it.  I thought it was a typo; my bad.

#### Tags Extraction and Validation

The Verifier chooses tags from the selected CoRIMs—including CoMID, CoSWID, CoBOM, and CoTS.

The Verifier MUST discard all tags which are not syntactically and semantically valid.
In particular, any cross-referenced triples (e.g., CoMID-CoSWID linking triples and CoBOM-listed tags) MUST be successfully resolved.

#### Tag activation by CoBOM

This step can be skipped of all tags are implicitly active.

CoBOMs which are not within their validity period are discarded.
The Verifier processes all CoBOMs that are valid at the point in time of Evidence Appraisal.

A Verifier MAY decide to discard some of the available and valid CoBOMs depending on any locally configured authorization policies.
(Such policies model the trust relationships between the Verifier Owner and the relevant suppliers, and are out of the scope of the present document.)
For example, a composite device ({{Section 3.3 of -rats-arch}}) is likely to be fully described by multiple CoRIMs, each signed by a different supplier.
In such case, the Verifier Owner may instruct the Verifier to discard tags activated by supplier CoBOMs that are not also activated by the trusted integrator.

> If the Verifier Policy requires CoBOMs, then after the Verifier has processed all CoBOMs it MUST discard any tags which have not been activated by a CoBOM.

Reviewer 2: Removal of sec-ev-coll goes too far. This section corresponds to the CoRIM Selection, only for Evidence. I suggest:
```
#### Evidence Selection

All available Evidence in the Appraisal Session's inputs are checked for validity.

Evidence that is not within it's validity period, or that cannot be associated with an authenticated and authorized source MUST be discarded.

Evidence that has been secured by a cryptographic mechanism, such as a signature, that fails validation MUST be discarded.

Evidence selection MUST yield at least one usable entry.

Selected Evidence is transformed into an internal representation (see {{sec-phase1-trans}}).
```
The new text doesn't seem to belong in this section.

Author: Added.

#### Cryptographic Validation of Evidence {#sec-crypto-validate-evidence}

> The collected Evidence is transformed to an internal representation, making it suitable for appraisal processing.
> 

Reviewer 2: ```suggestion

```

Evidence transformation algorithms may be well-known, may be defined by a CoRIM profile ({{sec-corim-profile-types}}), or may be supplied dynamically.
The handling of dynamic Evidence transformation algorithms is out of scope for this document.

### Appraisal hermeticity

The Appraisal Context at the and of Phase 1 constitutes all inputs to the Appraisal Procedure.

> The same Appraisal Context processed at any time MUST produce the same Attestation Results.
> 
> The reason to lock the inputs before Attestation Appraisal is to ensure consistent ACS construction when accounting for any profile-driven semantics.

Author: Please help wordsmith this. There are obviously different levels of hermeticity and auditability that Verifier policy will impose. I don't want to require absolute determinism since that imposes a requirement on policy evaluation.
I don't like "same" since we're not prescribing any Attestation Result construction or representation, so there is no strong notion of sameness.

If we go with 
```suggestion
The reason to lock the inputs before Attestation Appraisal is for all Appraisal Procedure dependencies to be accounted for before interpreting them.                                                                                            
Fully determined inputs SHOULD ensure consistent ACS construction when accounting for any profile-driven semantics.                                                                                                                             
```

Reviewer: I agree that the old L1906 is not 100% correct: it should probably have said instead:

> The same Appraisal Context processed at any time MUST produce the same ACS.

I like the text you suggest.  The only thing is we should state more strongly that what we are trying to achieve is _deterministic ACS construction._

Reviewer 2: > Fully determined inputs SHOULD ensure consistent ACS construction when accounting for any profile-driven semantics.

```suggestion
Given the same Appraisal Context, different Verifier appraisals MUST produce deterministic results.

Verifier inputs are locked (see Step 1) before Attestation Appraisal to ensure ACS construction is deterministic.
```
Text above addresses the question of profile or policy inputs rendering non-deterministic ACS results.

Evidence transformation algorithms may be well-known, defined by a CoRIM profile ({{sec-corim-profile-types}}), or supplied dynamically.
The handling of dynamic Evidence transformation algorithms is out of scope for this document.

### Appraisal hermeticity

The Appraisal Context at the and of Phase 1 constitutes all inputs to the Appraisal Procedure.

> Given the same Appraisol Context, different Verifier appraisals MUST produce deterministic results for phases 2, 3, and 4.

Reviewer 3: ```suggestion
Given the same Appraisal Context, different Verifier appraisals MUST produce deterministic results for phases 2, 3, and 4.
```

### Reference Values Corroboration and Augmentation (Phase 3) {#sec-phase3}

> Reference Value Providers (RVP) publish Reference Values using the Reference Values Triple ({{sec-comid-triple-refval}}) which are transformed ({{sec-ref-trans}}) into an internal representation ({{sec-ir-ref-val}}) of `rv` relations of `rv` relations.

Reviewer 2: ```suggestion
Reference Value Providers (RVP) publish Reference Values using the Reference Values Triple ({{sec-comid-triple-refval}}) which are transformed ({{sec-ref-trans}}) into an internal representation ({{sec-ir-ref-val}}) of `rv` relations.
```
